### PR TITLE
Group-by aggregation based optimization for UNBOUNDED `collect_set` window function

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes https://github.com/NVIDIA/spark-rapids/issues/10114.

This change introduces an optimization for the `collect_set()` window function, when run as an fully `UNBOUNDED` window function, as follows:

```sql
SELECT COLLECT_SET(foo) OVER (PARTITION BY grp 
                              ORDER BY ord 
                              ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) ... 
```

It might be possible to solve most UNBOUNDED window aggregations via group-wise aggregations. This change addresses `collect_set` first.

This change currently builds on the changes @revans2 made over at #10158, and includes those changes here.  

A new window-exec is introduced specifically to address fully UNBOUNDED window functions.  Each input batch is processed via group-wise aggregations.  For groups that are known to be complete, the output is returned.  The results for incomplete groups are cached, and finalized only when the end of the group is encountered.

### CODE REVIEW

**Comment by mythrocks** [comment]
The operation seems to be functional, from my testing.  A few tasks remain:
1. Handling retries in the second stage.
2. Rooting out memory leaks.

I'm working on these now.

**Review by revans2 — COMMENTED** [review]
In general it looks good. I did a quick pass, but nothing in depth. Like you said it is very specific to collect_set, which is fine for now, but we should have some follow on work to figure out how to make it better.

**Comment by mythrocks** [comment]
I'm now trying to sort out the bugs revealed in the test I added.  I'll update when I have more.

**Comment by mythrocks** [comment]
A quick update: The code is close to complete.  There seems to be a memory leak that's proving hard to find.

**Comment by mythrocks** [comment]
I've resolved the memory leaks at last. @revans2, could you take another peek at this?

**Review by revans2 — COMMENTED** [review]
Sorry I am getting compile errors and a bunch of other odd things. Could you please upmerge to the latest in https://github.com/NVIDIA/spark-rapids/pull/10158

There have been some big changes. That or could you review it so we can merge it in, and then you can uperge to the latest code.

**Comment by mythrocks** [comment]
Sorry, I thought I'd upmerged. 

Yes, I should resolve these before we merge.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuWindowExecMeta.scala [hunk 1]
```diff
@@ -458,7 +462,26 @@ object GpuWindowExecMeta {
...
+        func match {
+          case _: GpuUnboundedToUnboundedWindowAgg => true
+          case GpuAggregateExpression(_: GpuUnboundedToUnboundedWindowAgg, _, _, _, _) => true
+          case _ => false
+        }
+      }
+    }
+  }
+
+  // TODO delete this if we can support min, max, and count with
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuWindowExecMeta.scala:482]: Thank you. After I've understood out how pre/post process are to be done for the other aggregations, I can get to enabling this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+    // thrown away. If we ever dedupe this with a COUNT(1) that already exists, then
+    // we need to update the output of this to have a clean way to say what is the count,
+    // and if that count is needed see repeatOps
+
+    val aggregationsOutput = windowAggs.map(_.toAttribute)
+    val boundAggregations = GpuBindReferences.bindGpuReferences(windowAggs, childOutput)
+    (aggregationsOutput, boundAggregations)
+  }
+
+  private def repeatOps(aggregationsOutput: Seq[Attribute])
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala:661]: Sorry for missing that. I'll fix this shortly.

File: tests/src/test/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowSuite.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      AttributeReference("_count", LongType, true)())
+
+    val rideAlongOutput = Seq(AttributeReference("a", ShortType, true)())
+    val (boundCount, repeatOutput, boundRepeat) =
+      GpuUnboundedToUnboundedAggWindowIterator.repeatOps(aggOutput)
+
+    val finalProject = GpuUnboundedToUnboundedAggWindowIterator.computeFinalProject(
+      rideAlongOutput, repeatOutput, repeatOutput ++ rideAlongOutput)
+
+    val conf = GpuUnboundedToUnboundedAggStages(Seq.empty, Seq.empty,
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply tests/src/test/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowSuite.scala:41]: Sorry I missed this. Fixing now.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuWindowExecMeta.scala [hunk 4]
```diff
@@ -501,7 +528,12 @@ object GpuWindowExecMeta {
         throw new IllegalArgumentException(
           s"Found unexpected expression $other in window exec ${other.getClass}")
     }
-    BatchedOps(running.toSeq, doublePass.toSeq, batchedBounded.toSeq, passThrough.toSeq)
+    // TODO: CALEB:  Reclassify `collect_set()` here. Separate window-exec.
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuWindowExecMeta.scala]: That was a stale code-marker. I've removed it. Thank you for pointing it out.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+    rideAlongColumns.close()
+    aggResult.close()
+  }
+}
+
+/**
+ * Utilities for conversion between SpillableColumnarBatch, ColumnarBatch, and cudf.Table.
+ */
+object TableAndBatchUtils {
+  def toSpillableBatch(cb: ColumnarBatch): SpillableColumnarBatch = {
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala]: I've added documentation for the lot.  
The `toSpillableBatch()` function that adopts the argument has been renamed to `adoptAndMakeSpillable`.

I think this clarifies, but I'd be happy to make further improvements.

File: tests/src/test/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowSuite.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mythrocks) [reply tests/src/test/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowSuite.scala]: Should be fixed now.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+    }
+    withResource(fixedCols) { _ =>
+      new cudf.Table(fixedCols: _*)
+    }
+  }
+
+  // Append column at the end to account for the added `GpuCount(1)`.
+  private def preProcess(inputCB: ColumnarBatch): ColumnarBatch = {
+    withResource(GpuColumnVector.from(inputCB)) { inputTable =>
+      withResource(cudf.Scalar.fromInt(1)) { one =>
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala:194]: I've raised https://github.com/NVIDIA/spark-rapids/issues/10374. I'll get on it after this PR.

**Comment by revans2** [comment]
Still getting errors on compile. Looks like some style issues with the imports.

**Comment by revans2** [comment]
I did some quick scale tests and it looks really good. I'll try to look through all of the code on Monday.

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
Build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala [hunk 7]
```diff
@@ -0,0 +1,1172 @@
...
+        }
+      }
+      val result = currIter.next()
+      subIterator = Some(currIter)
+      result
+    }
+  }
+}
+
+case class PartitionedFirstPassAggResult(firstPassAggResult: FirstPassAggResult,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala:264]: nit: I don't think this really fits with being a `case class`. We might want to change it to be just a class.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala [hunk 8]
```diff
@@ -0,0 +1,1172 @@
...
+  var lastGroupAggResult: Option[SpillableColumnarBatch] = None
+  var lastGroupRideAlong: Option[SpillableColumnarBatch] = None
+  var otherGroupAggResult: Option[SpillableColumnarBatch] = None
+  var otherGroupRideAlong: Option[SpillableColumnarBatch] = None
+
+  val numGroups: Int = firstPassAggResult.aggResult.numRows()
+  private val numGroupingKeys: Int = boundStages.boundPartitionSpec.size
+  private val numRideAlongRows: Int = firstPassAggResult.rideAlongColumns.numRows()
+
+  if (numGroups < 2) {
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala:275]: I should have made this clearer in the calling code.  This shouldn't be a problem.

Here's the call-site, with context.
```scala
    if (newData.aggResult.numRows() == 1) {
      // All the aggregation results are for the same group.
      // Add the lot to "incomplete".  No results with this input.
      rideAlongColumnsPendingCompletion.add(newData.rideAlongColumns.incRefCount())
      aggResultsPendingCompletion += newData.aggResult.incRefCount()
      None
    }
    else {
      opTime.ns {
        val partitioned = withRetryNoSplit(newData) {
          PartitionedFirstPassAggResult(_, boundStages)
        } 
      // ...
    }
```

At [line#451](https://github.com/NVIDIA/spark-rapids/pull/10248/files#diff-906dc93a262624c8988ffd3e84ec3d861f842be2924809a0d5784599c16d0b3dR451), we check the number of groups.  If there's only a single group, the entire column is marked pending-completion. `PartitionedFirstPassAggResult` isn't constructed unless there is evidence of a group ending, i.e. at least two groups.

I can try make that clearer in the code.

[the remark this answers is not in the document]
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala:286]: I've added documentation for `PartitionedFirstPassAggResult`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala (hunk 7 above)
Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala]: Thanks for pointing that out, @revans2.  That helped me better understand the use of `case class`.  I've changed this to a regular class now.

**Comment by mythrocks** [comment]
CI still seems to be failing because of https://github.com/NVIDIA/spark-rapids/issues/10382. I'll retry the build after #10382 is resolved.

**Comment by mythrocks** [comment]
Build

**Review by revans2 — APPROVED** [review]

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
@revans2: I've rebased this change since #10158 was merged.  The tests are passing, thankfully.

**Review by revans2 — APPROVED** [review]

**Pull request merged**